lang2/macro/src/ir/into_hir.rs

```diff
@@ -0,0 +1,656 @@
+                bail!(
+                    item_mod,
+                    "contract module must be inline, e.g. `mod m {{ ... }}`",
+                )
+            }
+            Some((_brace, items)) => items.clone(),
+        };
+        use itertools::Itertools as _;
+        let (ink_items, rust_items): (Vec<_>, Vec<_>) = items
+            .into_iter()
+            .map(ir::Item::try_from)
+            .collect::<Result<Vec<_>>>()?
+            .into_iter()
+            .partition_map(|item| {
+                match item {
+                    ir::Item::Ink(ink_item) => Either::Left(ink_item),
+                    ir::Item::Rust(rust_item) => Either::Right(rust_item),
+                }
+            });
+        let (storage, events, functions) = split_items(ink_items)?;
+        if functions.iter().filter(|f| f.is_constructor()).count() == 0 {
+            bail!(
+                &item_mod,
+                "ink! contracts require at least one `#[ink(constructor)]`"
+            )
+        }
+        if functions.iter().filter(|f| f.is_message()).count() == 0 {
+            bail!(
+                &item_mod,
+                "ink! contracts require at least one `#[ink(message)]`"
+            )
+        }
+        let meta_info = ir::MetaInfo::try_from(params)?;
+        Ok(Self {
+            mod_token: item_mod.mod_token,
+            ident: item_mod.ident,
+            attrs: item_mod.attrs,
+            meta_info,
+            storage,
+            events,
+            functions,
+            non_ink_items: rust_items,
+        })
+    }
+}
+
+impl TryFrom<ir::Params> for ir::MetaInfo {
+    type Error = syn::Error;
+
+    fn try_from(params: ir::Params) -> Result<Self> {
+        let mut unique_params = HashSet::new();
+        let mut env_types = None;
+        let mut ink_version = None;
+        let mut dynamic_allocations: Option<bool> = None;
+        let mut compile_as_dependency: Option<bool> = None;
+        for param in params.params.iter().cloned() {
+            let name = param.ident().to_string();
+            if !unique_params.insert(name) {
+                bail_span!(param.span(), "encountered parameter multiple times",)
+            }
+            match param {
+                ir::MetaParam::Types(param) => {
+                    env_types = Some(ir::MetaTypes::try_from(param)?)
+                }
+                ir::MetaParam::Version(param) => ink_version = Some(param.data),
+                ir::MetaParam::DynamicAllocations(param) => {
+                    dynamic_allocations = Some(param.value.value)
+                }
+                ir::MetaParam::CompileAsDependency(param) => {
+                    compile_as_dependency = Some(param.value.value)
+                }
+            }
+        }
+        let ink_version = match ink_version {
+            None => {
+                bail_span!(
+                    params.span(),
+                    "expected `types` argument at `#[ink::contract(..)]`",
```

Comment:
Shouldn't this be the following?
```suggestion
                    "expected `version` argument at `#[ink::contract(..)]`",
```